api/envoy/config/common/matcher/v3/matcher.proto

```diff
@@ -16,6 +18,126 @@ option (udpa.annotations.file_status).package_version_status = ACTIVE;
 
 // [#protodoc-title: Unified Matcher API]
 
+// A matcher, which may traverse a matching tree in order to result in a match action.
+// During matching, the tree will be traversed until a match is found, or if no match
+// is found the action specified by the most specific on_no_match will be evaluated.
+// As an on_no_match might result in another matching tree being evaluated, this process
+// might repeat several times until the final OnMatch (or no match) is decided.
+message Matcher {
+  // What to do if a match is successful.
+  message OnMatch {
+    oneof on_match {
+      option (validate.required) = true;
+
+      // Nested matcher to evaluate.
+      // If the nested matcher does not match and does not specify
+      // on_no_match, then this matcher is considered not to have
+      // matched, even if a predicate at this level or above returned
+      // true.
+      Matcher matcher = 1;
+
+      // Protocol-specific action to take.
+      core.v3.TypedExtensionConfig action = 2;
+    }
+  }
+
+  // A linear list of field matchers.
+  // The field matchers are evaluated in order, and the first match
+  // wins.
+  message MatcherList {
+    // Predicate to determine if a match is successful.
+    message Predicate {
+      // Predicate for a single input field.
+      message SinglePredicate {
+        // Protocol-specific specification of input field to match on.
+        core.v3.TypedExtensionConfig input = 1 [(validate.rules).message = {required: true}];
+
+        oneof matcher {
+          option (validate.required) = true;
+
+          // Use existing infrastructure for actually matching the
+          // value.
+          type.matcher.v3.ValueMatcher value_match = 2;
+
+          // Extension for custom matching logic.
+          core.v3.TypedExtensionConfig custom_match = 3;
+        }
+      }
+
+      // A list of matchers. Used to allow using a list within a oneof.
+      message PredicateList {
+        repeated Predicate predicate = 1 [(validate.rules).repeated = {min_items: 1}];
+      }
+
+      oneof match_type {
+        option (validate.required) = true;
+
+        // A single predicate to evaluate.
+        SinglePredicate single_predicate = 1;
```

Comment:
Is there a semantic difference between `SinglePredicate` and either of `PredicateList` with single element?
Have you considered either removing `single_predicate` or setting `min_items` to 2? (Otherwise there are 3 different ways to specify single predicate).

Reply:
I'd be fine with setting `min_items` to 2 for `or_matcher` and `and_matcher`.

Reply:
+1

Reply:
Why have SinglePredice? It seems redundant.

Reply:
How is `SinglePredicate` redundant?  It looks like that's the only way to encode a specific input value to be matched against a specific matcher.  What would you suggest instead?

Reply:
Use PredicatList with either Or or And semantics. It provides the exact same behavior.

Reply:
But inside the `PredicateList`, the only way to actually specify an individual predicate is `SinglePredicate`.  If we remove that, how will anyone be able to put anything inside of `PredicateList`?

Reply:
I see. Would it be clearer if we name this `LeafPredicate`?

Reply:
I was thinking that `SinglePredicate` made more sense in the simple case where only one predicate is actually needed.  But I could also live with `LeafPredicate` if others like that better.

Reply:
I don't mind either way that much, it is a bit confusing as is given this thread as evidence, so whatever the consensus on understandability is.

Reply:
I don't have a strong opinion here, I don't think the specific adjective we use here changes much. I think I slightly prefer `SinglePredicate` because I think it would make more sense for someone that didn't grok the fact that this AND/OR scheme models a tree.

Reply:
I'll merge this later today as is unless anyone has any strong objections